### PR TITLE
Telophase Keyboard fixes and keymap. 

### PR DESCRIPTION
**Pull request by winterscar**
## Description
The telophase now comes in two hardware versions. I've pulled some changes from southpawengineer's fork of qmk into here, and added some changes to support these hardware changes

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/telophase/config.h [hunk 1]
```diff
@@ -25,13 +25,13 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-#define DESCRIPTION     q.m.k. keyboard firmware for Telophase
+#define MANUFACTURER    SouthpawDesign
+#define PRODUCT         telophase
+#define DESCRIPTION     q.m.k. keyboard firmware for TELOPHASE
 
 /* key matrix size */
-#define MATRIX_ROWS 4
-#define MATRIX_COLS 12
+#define MATRIX_ROWS 5
+#define MATRIX_COLS 14
```
Comment (fauxpark) [inline keyboards/telophase/config.h]: Is this correct?

File: keyboards/telophase/config.h [hunk 2]
```diff
@@ -41,11 +41,19 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 #define ONESHOT_TIMEOUT 500
 
+
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
+
```
Comment (fauxpark) [inline keyboards/telophase/config.h]: This isn't necessary, as it's the default in QMK core.

File: keyboards/telophase/config.h [hunk 3]
```diff
@@ -41,11 +41,19 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 /*
  * Feature disable options
  *  These options are also useful to firmware size reduction.
  */
 
+#define PREVENT_STUCK_MODIFIERS
+
```
Comment (fauxpark) [inline keyboards/telophase/config.h]: Please remove this, it doesn't do anything anymore.

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]:
```suggestion
```

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    _______, _______, MU_MOD,  AU_ON,   AU_OFF,  AG_NORM, AG_SWAP, _______, _______, _______,  _______, _______,
+    _______, MUV_DE,  MUV_IN,  MU_ON,   MU_OFF,  MI_ON,   MI_OFF,  TERM_ON, TERM_OFF, _______, _______, _______,
+    _______, _______, _______, _______, _______, _______, _______, _______, _______,  _______, _______, _______
+)
+
+};
+
+void matrix_init_user() {
+  return;
+}
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]: This function can be removed; it doesn't do anything.

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+uint32_t layer_state_set_user(uint32_t state) {
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]:
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
```

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    /// TEXT MACROS
+    case M_EADDR:
+      if (record->event.pressed) {
+        // when keycode QMKBEST is pressed
+        SEND_STRING("your@email.here" SS_DOWN(X_LCTRL) SS_TAP(X_LEFT) SS_TAP(X_LEFT) SS_TAP(X_LEFT) SS_UP(X_LCTRL) SS_TAP(X_LEFT));
+      }
+      break;
+    case M_RGRDS:
+      if (record->event.pressed) {
+        SEND_STRING("Regards," SS_TAP(X_ENTER) "Mat.");
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]:
```suggestion
        SEND_STRING("Regards,\nMat.");
```

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      }
+      break;
+    case M_GITPS:
+      if (record->event.pressed) {
+        SEND_STRING("git add ." SS_TAP(X_ENTER) "git commit -m 'I made some changes'" SS_TAP(X_ENTER) "git push");
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]:
```suggestion
        SEND_STRING("git add .; git commit -m 'I made some changes'\ngit push");
```

File: keyboards/telophase/keymaps/mathieu/keymap.c [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      }
+      break;
+  }
+  return true;
+}
+
+void matrix_scan_user(void) {
+  return;
+}
```
Comment (fauxpark) [inline keyboards/telophase/keymaps/mathieu/keymap.c]: This doesn't do anything either.

File: keyboards/telophase/readme.md [hunk 10]
```diff
@@ -11,6 +11,10 @@ Make example for this keyboard (after setting up your build environment):
 
 See [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) then the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information.
 
-## Mitosis Notes
+## Hardware versions
+
+There are two versions of the pro micro, a 3v and a 5v version. Depending on the version you're using, you'll need to set the CPU speed to `8000000` or `16000000` respectivly. Do this by changing the F_CPU version in rules.mk
```
Comment (fauxpark) [inline keyboards/telophase/readme.md]:
```suggestion
There are two versions of the Pro Micro, a 3V and a 5V version. Depending on the version you're using, you'll need to set the processor frequency to `8000000` or `16000000` respectively. Do this by changing `F_CPU` in rules.mk.
```

File: keyboards/telophase/rules.mk [hunk 11]
```diff
@@ -1,15 +1,64 @@
...
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
+
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: Please undo this change. All of this stuff is defined automatically by QMK, based on `MCU` and `BOOTLOADER`.

File: keyboards/telophase/rules.mk [hunk 12]
```diff
@@ -1,15 +1,64 @@
...
+OPT_DEFS += -DCATERINA_BOOTLOADER
+INTERPHASE_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
+                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
+
+# # project specific files
+SRC = matrix.c
+
+
 # MCU name
+#MCU = at90usb1287
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/telophase.h [hunk 13]
```diff
@@ -46,19 +48,17 @@
 // This a shortcut to help you visually see your layout.
 // The first section contains all of the arguements
 // The second converts the arguments into a two-dimensional array
-#define LAYOUT( \
+#define KEYMAP( \
```
Comment (fauxpark) [inline keyboards/telophase/telophase.h]: Please undo this change - the naming scheme for layout macros is `LAYOUT[_*]`.

File: keyboards/telophase/telophase.h [hunk 14]
```diff
@@ -2,6 +2,8 @@
 #define TELOPHASE_H
 
 #include "quantum.h"
+#include "matrix.h"
+#include <stddef.h>
```
Comment (fauxpark) [inline keyboards/telophase/telophase.h]: I don't think these includes are required here.

File: keyboards/telophase/telophase.h [hunk 15]
```diff
@@ -2,6 +2,8 @@
 #define TELOPHASE_H
```
Comment (fauxpark) [inline keyboards/telophase/telophase.h]: Could you please change this ifndef/define to just `#pragma once`?

File: keyboards/telophase/telophase.h [hunk 16]
```diff
@@ -46,19 +48,17 @@
...
+  { k11, k10, k09, k08, k07, k06, KC_NO, KC_NO, k05, k04, k03, k02, k01, k00},  \
+  { k23, k22, k21, k20, k19, k18, KC_NO, KC_NO, k17, k16, k15, k14, k13, k12},  \
+  { k35, k34, k33, k32, k31, k30, KC_NO, KC_NO, k29, k28, k27, k26, k25, k24},  \
+  { k47, k46, k45, k44, k43, k42, KC_NO, KC_NO, k41, k40, k39, k38, k37, k36},  \
+  { KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO}  \
 }
-
-#define LAYOUT_ortho_4x12 LAYOUT
-
 #endif
```
Comment (fauxpark) [inline keyboards/telophase/telophase.h]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 17]
```diff
@@ -1,15 +1,64 @@
+
+OPT_DEFS += -DTELOPHASE_PROMICRO
+OPT_DEFS += -DCATERINA_BOOTLOADER
+INTERPHASE_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
+                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: These aren't needed or used. QMK provides a way to flash keyboards intelligently based off the bootloader defined, with `make keyboard:keymap:flash`.

File: keyboards/telophase/rules.mk [hunk 18]
```diff
@@ -1,15 +1,64 @@
...
+
+# # project specific files
+SRC = matrix.c
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: This should be moved back to the end of the file.

File: keyboards/telophase/config.h (hunk 1 above)
Comment (winterscar) [reply keyboards/telophase/config.h]: This comes straight from the manufacturers fork of qmk. It works on my keyboard.

File: keyboards/telophase/rules.mk [hunk 19]
```diff
@@ -11,6 +8,24 @@ MCU = atmega32u4
...
+#     processor frequency in Hz. You can then use this symbol in your source code to
+#     calculate timings. Do NOT tack on a 'UL' at the end, this will be done
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: Please delete this block.

File: keyboards/telophase/rules.mk [hunk 20]
```diff
@@ -11,6 +8,24 @@ MCU = atmega32u4
 #   ATmega328P   USBasp
 BOOTLOADER = caterina
 
+# MCU Selection
+MCU = atmega32u4
+
+# Processor frequency.
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
# Processor frequency
```

File: keyboards/telophase/rules.mk [hunk 21]
```diff
@@ -11,6 +8,24 @@ MCU = atmega32u4
...
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+
+F_CPU = 8000000
+#F_CPU = 16000000
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 22]
```diff
@@ -11,6 +8,24 @@ MCU = atmega32u4
...
+#     calculate timings. Do NOT tack on a 'UL' at the end, this will be done
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+
+F_CPU = 8000000
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: And this whole thing needs to go at the top of the file.

File: keyboards/telophase/rules.mk [hunk 23]
```diff
@@ -11,6 +8,24 @@ MCU = atmega32u4
 #   ATmega328P   USBasp
 BOOTLOADER = caterina
 
+# MCU Selection
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
# MCU name
```

File: keyboards/telophase/rules.mk [hunk 24]
```diff
@@ -27,15 +42,11 @@ NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
...
+
 USB = /dev/ttyACM0
 
 # upload: build
-# 	$(MITOSIS_UPLOAD_COMMAND)
-
-OPT_DEFS += -DMITOSIS_PROMICRO
-MITOSIS_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
-                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
+# 	$(TELOPHASE_UPLOAD_COMMAND)
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 25]
```diff
@@ -27,15 +42,11 @@ NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
 # MIDI_ENABLE = YES 		# MIDI controls
 UNICODE_ENABLE = YES 		# Unicode
 # BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
-LAYOUTS = ortho_4x12
+
 USB = /dev/ttyACM0
 
 # upload: build
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 26]
```diff
@@ -27,15 +42,11 @@ NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
 # MIDI_ENABLE = YES 		# MIDI controls
 UNICODE_ENABLE = YES 		# Unicode
 # BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
-LAYOUTS = ortho_4x12
+
 USB = /dev/ttyACM0
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/config.h (hunk 1 above)
Comment (fauxpark) [reply keyboards/telophase/config.h]: According to this: https://southpawdesign.net/products/wireless-split-keyboard-group-buy
the Telophase is actually 4x12, and it's the Helicase that is 5x14.

File: keyboards/telophase/telophase.h [hunk 27]
```diff
@@ -53,12 +52,9 @@
...
-	{ k24, k25, k26, k27, k28, k29,      k30, k31, k32, k33, k34, k35 }, \
-	{ k36, k37, k38, k39, k40, k41,      k42, k43, k44, k45, k46, k47 }, \
+  { k11, k10, k09, k08, k07, k06, KC_NO, KC_NO, k05, k04, k03, k02, k01, k00},  \
+  { k23, k22, k21, k20, k19, k18, KC_NO, KC_NO, k17, k16, k15, k14, k13, k12},  \
+  { k35, k34, k33, k32, k31, k30, KC_NO, KC_NO, k29, k28, k27, k26, k25, k24},  \
+  { k47, k46, k45, k44, k43, k42, KC_NO, KC_NO, k41, k40, k39, k38, k37, k36},  \
+  { KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO}  \
 }
-
-#define LAYOUT_ortho_4x12 LAYOUT
```
Comment (fauxpark) [inline keyboards/telophase/telophase.h]: Please put this back.

File: keyboards/telophase/telophase.h [hunk 28]
```diff
@@ -46,15 +46,16 @@
...
+  k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, \
+  k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34, k35, \
+  k36, k37, k38, k39, k40, k41, k42, k43, k44, k45, k46, k47  \
 ) \
 { \
-  { k11, k10, k09, k08, k07, k06, KC_NO, KC_NO, k05, k04, k03, k02, k01, k00},  \
-  { k23, k22, k21, k20, k19, k18, KC_NO, KC_NO, k17, k16, k15, k14, k13, k12},  \
-  { k35, k34, k33, k32, k31, k30, KC_NO, KC_NO, k29, k28, k27, k26, k25, k24},  \
-  { k47, k46, k45, k44, k43, k42, KC_NO, KC_NO, k41, k40, k39, k38, k37, k36},  \
-  { KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO}  \
```
Comment (winterscar) [inline keyboards/telophase/telophase.h]: These changes were specific to my use-case as I wired my keyboard up backwards. Is there an appropriate place to put my specific `LAYOUT` macro? somewhere in the `/mathieu` folder?

File: keyboards/telophase/rules.mk [hunk 29]
```diff
@@ -27,15 +32,6 @@ NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
...
-# upload: build
-# 	$(MITOSIS_UPLOAD_COMMAND)
-
-OPT_DEFS += -DMITOSIS_PROMICRO
-MITOSIS_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
-                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
 
 # # project specific files
-SRC = matrix.c
+SRC = matrix.c
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
SRC = matrix.c

LAYOUTS = ortho_4x12
```

File: keyboards/telophase/rules.mk [hunk 30]
```diff
@@ -1,6 +1,10 @@
 # MCU name
 MCU = atmega32u4
 
+# Processor frequency
+F_CPU = 8000000
+
+
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 31]
```diff
@@ -11,6 +15,7 @@ MCU = atmega32u4
 #   ATmega328P   USBasp
 BOOTLOADER = caterina
 
+
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]:
```suggestion
```

File: keyboards/telophase/rules.mk [hunk 32]
```diff
@@ -1,6 +1,10 @@
 # MCU name
 MCU = atmega32u4
 
+# Processor frequency
+F_CPU = 8000000
+
```
Comment (fauxpark) [inline keyboards/telophase/rules.mk]: If this is using a Pro Micro, I think this should actually be removed, as the 16MHz variant is far more common. Users building one of these with an 8MHz Pro Micro can just add it to their keymap's rules.mk.

File: keyboards/telophase/telophase.h (hunk 28 above)
Comment (fauxpark) [reply keyboards/telophase/telophase.h]: I believe you can just add it to telophase.h as `LAYOUT_reverse` or something similar.
